src/transformers/models/qwen2_audio/modeling_qwen2_audio.py

```diff
@@ -0,0 +1,1239 @@
+            )
+
+        output_attentions = output_attentions if output_attentions is not None else self.config.output_attentions
+        output_hidden_states = (
+            output_hidden_states if output_hidden_states is not None else self.config.output_hidden_states
+        )
+        return_dict = return_dict if return_dict is not None else self.config.use_return_dict
+        input_features = input_features.to(dtype=self.conv1.weight.dtype, device=self.conv1.weight.device)
+        inputs_embeds = nn.functional.gelu(self.conv1(input_features))
+        inputs_embeds = nn.functional.gelu(self.conv2(inputs_embeds))
+
+        inputs_embeds = inputs_embeds.permute(0, 2, 1)
+        embed_pos = self.embed_positions.weight
+
+        hidden_states = inputs_embeds + embed_pos
+        hidden_states = nn.functional.dropout(hidden_states, p=self.dropout, training=self.training)
+
+        encoder_states = () if output_hidden_states else None
+        all_attentions = () if output_attentions else None
+
+        # check if head_mask has a correct number of layers specified if desired
+        if head_mask is not None:
+            assert head_mask.size()[0] == (
+                len(self.layers)
+            ), f"The head_mask should be specified for {len(self.layers)} layers, but it is for {head_mask.size()[0]}."
+
+        for idx, encoder_layer in enumerate(self.layers):
+            if output_hidden_states:
+                encoder_states = encoder_states + (hidden_states,)
+            # add LayerDrop (see https://arxiv.org/abs/1909.11556 for description)
+            to_drop = False
+            if self.training:
+                dropout_probability = torch.rand([])
+                if dropout_probability < self.layerdrop:  # skip the layer
+                    to_drop = True
+
+            if to_drop:
+                layer_outputs = (None, None)
+            else:
+                if self.gradient_checkpointing and self.training:
+                    layer_outputs = self._gradient_checkpointing_func(
+                        encoder_layer.__call__,
+                        hidden_states,
+                        attention_mask,
+                        (head_mask[idx] if head_mask is not None else None),
+                        output_attentions,
+                    )
+                else:
+                    layer_outputs = encoder_layer(
+                        hidden_states,
+                        attention_mask,
+                        layer_head_mask=(head_mask[idx] if head_mask is not None else None),
+                        output_attentions=output_attentions,
+                    )
+
+                hidden_states = layer_outputs[0]
+
+            if output_attentions:
+                all_attentions = all_attentions + (layer_outputs[1],)
+
+        hidden_states = hidden_states.permute(0, 2, 1)
+        hidden_states = self.avg_pooler(hidden_states)
+        hidden_states = hidden_states.permute(0, 2, 1)
```

Comment:
Besides this particular snippet, this corresponds almost entirely to Whisper implementation, right?

Small questions on my side, are you finetuning the Whisper weights? If so, I wonder if it'd make sense to use `WhisperEncoder` directly, since it'd make the implementation much more easier to read and to use!

I think it would make sense to either:
* Add a config parameter to WhisperConfig which allows [to not use a `layer_norm` in the Encoder](https://github.com/faychu/transformers/blob/066146fd43f07f024de63e17a5bbda4a190652c1/src/transformers/models/whisper/modeling_whisper.py#L1078), and thus to rewrite the `Qwen2AudioEncoderModel` with 3 objects: 1. an `audio_encoder`, i.e an AutoModel that loads WhisperEncoder. 2. `avg_pooler` 3. `layer_norm`.
* Or a more straightforward method: add the possibility to add an `avg_pooler` [here](https://github.com/faychu/transformers/blob/066146fd43f07f024de63e17a5bbda4a190652c1/src/transformers/models/whisper/modeling_whisper.py#L1078) to the `WhisperEncoder` code, and use it instead of `Qwen2AudioEncoderModel`.

Reply:
Thanks for your suggestions. Yes, I finetune the Whisper weights. Enable whisper encoder to accept any chunk length

Reply:
Thanks for your suggestions.

Reply:
We sincerely appreciate your suggestion! In the future, we may update the audio encoder to support arbitrary-length chunks and potentially include more types of position embeddings. To avoid any possible impact on the Whisper model branch, we would prefer to keep `Qwen2AudioEncoder`.

Reply:
Thanks for the explanation! 

Given this new information, I think it would make sense to keep the code as it is. However, I'd still be in favor of making it clearer that it's mostly the Whisper encoder model.

To do so, you can add "# Copied from..." and "# Ignore copy" strategically:
1. in the init and in the forward
2. And add "# Ignore copy" where relevant, i.e where using `avg_pooler`


Also cc @amyeroberts and @sanchit-gandhi for a second opinion on this thread

Reply:
I'd be pro this suggestion: # Copied from..." and "# Ignore copy" strategically.

Reply:
Thank you, I will add these as you suggested.

Reply:
already added `# Copied from ...` and `# Ignore copy` here.